### PR TITLE
Code mode: validate a script without executing it

### LINKED ISSUE #647: Code mode: no way to validate a script without executing it
## Problem

There is no way to check a code-mode script without executing it.

The agent authors JavaScript blind against the `servers.*` stubs built from the scoped catalog (`script_catalog_tools`, wired at `src-tauri/src/bin/toolport-gateway.rs:4380-4390`). The first syntax error, misspelled tool name, or wrong argument shape is only discovered by running the script — which costs a full round trip, and may have already committed side effects on the calls that succeeded before the bad one.

That is the wrong shape for an authoring loop. It makes every draft a commitment.

## Proposed fix

Accept an optional `validate: true` argument on `toolport_run_script`. In that mode:

1. Compile the script (catches syntax errors).
2. Bind the same `servers.*` typed stubs from the caller's scoped catalog (catches unknown server/tool names and out-of-scope calls).
3. Swap the call binding for one that **records** intended `{name, args}` and returns a stub value instead of re-entering `execute_call`. Nothing downstream executes.
4. Return the recorded call plan, so the agent can inspect what the script *would* do.

Resource limits still apply, so a validate run cannot wedge the gateway either.

Note this is a linting aid, not a security control — code mode is explicitly not a security boundary (`src-tauri/src/codemode.rs:9-12`), and a validate pass cannot predict a script whose call sequence depends on real returned data. The value is catching the large class of mistakes that are static.

## Why now

Prerequisite for persistent routines (#625): saving a routine that was never validated persists a landmine that misfires later, unattended. Valuable on its own for one-off scripts too.

### PR DESCRIPTION
**Pull request by btsouth**
Fixes #647. Follows #661 (the executed-call ledger).

The agent authored scripts blind — the first syntax error, misspelled tool
name, or out-of-scope server was only discovered by running the script, which
costs a round trip and may already have committed side effects. Every draft was
a commitment.

`toolport_run_script` now accepts `validate: true`: compile, bind the same
client-scoped `servers.*` surface, enforce the same limits, return the plan of
calls it *would* make, execute none.

### Why this is low-risk

**`run_script` is unchanged.** The call binding is already an injected
parameter, so validate mode is just a recording closure passed in its place.
Nothing in the sandbox execution path — including the code #661 just touched —
moved.

**Nothing downstream can execute, structurally.** `validate_script_dispatch`
takes no router and no registry, and dispatch branches to it before the
executing binding is built. There is no path to `execute_call` to get wrong.

### Design decisions

**The plan includes arguments; the #646 ledger does not.** Not an
inconsistency. In validate mode nothing downstream runs, so a recorded argument
can only have come from the script or the `data` the agent passed in — both
already the caller's own. No downstream payload can reach the plan.

**`complete` is the load-bearing field.** Validation calls return an empty
result envelope, so a script that reads fields off a result diverges there. A
partial plan presented as a whole plan is what would make validation worse than
none, so an incomplete run says so in the structured output *and* the text, and
names "branched on a value a dry run cannot supply" alongside "real defect" as
causes. This is the same lesson as the two reviewer findings on #661: a
misleading report is worse than an absent one.

**No `savings::record_orchestration`.** A dry run replaced no round-trips;
counting it would inflate the metric the feature is judged by.

**`fetchResult` is stubbed, not unbound.** An unbound binding fails closed and
would report a validation failure for a script that pages results correctly.

Not a security control — code mode is explicitly not a security boundary, and
this is documented in the function and the tool description.

### Testing

- [x] `npm run test:rust:windows` — 783 lib + 233 gateway (up 4) + 29 others
- [x] Plan shape and indexes, syntax error, partial-plan reporting, out-of-scope tool

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src-tauri/src/bin/toolport-gateway.rs [hunk 1]
```diff
@@ -4374,6 +4378,102 @@ fn script_catalog_tools(
...
+    allowed: Option<&std::collections::HashSet<String>>,
+    client: Option<&str>,
+) -> Value {
+    let plan = Arc::new(Mutex::new(Vec::<Value>::new()));
+
+    let recorder = Arc::clone(&plan);
+    let call: codemode::CallBinding = Arc::new(move |name: &str, args: Value| {
+        let mut plan = recorder.lock().unwrap_or_else(|e| e.into_inner());
+        let index = plan.len();
+        plan.push(json!({ "index": index, "name": name, "args": args }));
```
[the remark this answers is not in the document]
Comment (btsouth) [reply src-tauri/src/bin/toolport-gateway.rs]: Three of the five were real and are fixed; two are answered below.

**Name resolution — right, and it defeated the point of the feature.** The typed `servers.*` stubs reject unknown names on their own, but the string form reaches the binding directly, so a dry run planned calls `execute_call` would refuse. The recorder now checks each name against the same scoped catalog the stubs are built from, records `resolved` per entry, and fails validation naming every offender rather than stopping at the first. You also correctly spotted why the gap survived: my first test passed an empty catalog and still expected a plan. It uses a real catalog now.

**Overclaimed completeness — right, and the field name was the bug.** Renamed `complete` to `finished`, which claims only what it can prove: the run reached the end of the script. A guarded read of a stub result finishes happily with a plan a real run would not produce, so the success text and tool description both say so.

**Unchecked arguments — right, but a limit rather than a fix.** The crate has no JSON Schema validator, so this cannot do what `execute_call` does without a new dependency. Rather than leave a silent false pass, the limit is stated in the success text, the tool description, and the function docs, and a test pins it so changing it has to be deliberate.

**Shaping truncating the plan — not changing this one.** Unlike #661, the headline verdict lives in the text, which survives head-first truncation, and shaping appends a cursor marker telling the agent how to fetch the rest. The verdict is never lost and the detail stays reachable.

Mutation-checked that defeating the scope check fails two of the new tests.
